### DISCUSSION TITLE
How to I make a custom tag-order?

### DISCUSSION THREAD

**Discussion by vulns-are-features-too** [marlonrichert/zsh-autocomplete#155] [Q&A] [answered]
I'm trying to make use of my own `tag-order` with the following in my `.zshrc`

```zsh
zstyle ':completion:*' tag-order '! ancestor-directory recent-directories recent-files' 'flags arguments local-directories files builtins history-words'
```

With that setting,  `zstyle -L | grep 'tag-order'` outputs the following

```zsh
zstyle -e ':completion:*:complete:*' tag-order $'\n    if [[ CURRENT -ne 1 && $compstate[context] == (command|condition) &&\n        $PREFIX$SUFFIX == [-+]* ]]; then\n      reply=( "(|*-)argument-* (|*-)option[-+]* values" options - )\n    else\n      reply=( "! options *files *directories" "*files *directories" - )\n    fi\n  '
zstyle ':completion:*:expand:*' tag-order '! all-expansions original' -
zstyle ':completion:*:(approximate|correct):*' tag-order '! original' -
zstyle ':completion:*' tag-order '! ancestor-directory recent-directories recent-files' 'flags arguments local-directories files builtins history-words'
```

With my setting, I want to make suggestions appear in the order `flags -> arguments -> local-directories -> files -> builtins history-words` while removing entries for `ancestor-directory` and recents. However, zsh still suggests ancestors & recents and my local files are still at the bottom of my suggestions.

There's no clear documentation on this `tag-order` and I'm new to zsh so I'm not sure what exactly I'm supposed to do here. How should I make my `tag-order` work?

**Comment by marlonrichert** [comment, marked as answer]
First of all, despite what its name might suggest, `tag-order` doesn't govern the order in which completions are shown. Instead, it governs the order in which completions are _tried_ within a single completion function. Additionally, it can be used to disable certain completions, by removing them from `tag-order`.

If you want to change the order in which completions are _presented,_ then `group-order` is setting you'll need to use. Unlike `tag-order`, though, it supports neither `!` nor patterns, and it expects each group to be added as a separate value.

Additionally, for your setting to override the one defined by `zsh-autocomplete`, it needs to have more _weight._ This works similarly to CSS: The more elements there are in a selector, the more weight it has.

Finally, `flags` is neither a tag nor a group in Zsh's completion system. 🙂

What you want is something like this:
```zsh
zstyle ':completion:*:complete:*:' tag-order '! ancestor-directories recent-directories recent-files' -
zstyle ':completion:*:complete:*:' group-order \
  options arguments values local-directories files builtins history-words
```

Note the `-` at the end of `tag-order`. This tells it to not try any other tags. Without it, it would first try everything except the three tags listed with `!`, after which it would try all remaining tags.

Note also the extra `:` in each setting and the lack of a `*` at the end. This makes them have more weight than the ones defined by `zsh-autocomplete`.

I will see if I can improve the Readme on this.

**Reply by marlonrichert** [reply]
There: https://github.com/marlonrichert/zsh-autocomplete/blob/master/README.md#change-the-order-of-completions

Hopefully, it makes more sense now.

**Reply by L0laapk3** [reply]
Hi @marlonrichert,

Thanks for your amazing work!

This info appears to be no longer in the readme. Is it still accurate?

**Reply by Lockszmith-GH** [reply]
It looks that some overhaul of the README.md file was done back in 2021 and the section was removed on [this commit](https://github.com/marlonrichert/zsh-autocomplete/blob/b88c6abb58fcebfe993f60ff768b471b3705a2c2/README.md).

The [previous commit](https://github.com/marlonrichert/zsh-autocomplete/blob/4ec8790/README.md#change-the-order-of-completions) hold the old content.

<details><summary>how I found this:</summary>

```console
$ git clone https://github.com/marlonrichert/zsh-autocomplete.git
$ cd zsh-autocomplete
$ git log -S'### Change the order of completions' -- README.md
commit b88c6abb58fcebfe993f60ff768b471b3705a2c2
Author: Marlon Richert <marlonrichert@users.noreply.github.com>
Date:   Sun Mar 7 12:13:25 2021 +0200

    Rewrite "Settings" section in Readme

commit 3ecf30017642ecaccc72b23af48add3f7e7cac36
Author: Marlon Richert <marlonrichert@users.noreply.github.com>
Date:   Mon Dec 28 18:56:28 2020 +0200

    Improve `group-order` & add documentation

    Address discussion #155.

# Copy-paste the commit SHA, and list a couple of previous commits:
$ COMMIT=b88c6abb58fcebfe993f60ff768b471b3705a2c2; git log --oneline ${COMMIT}~3..${COMMIT}
b88c6ab Rewrite "Settings" section in Readme
4ec8790 Simplify "unambiguous" logic
204c263 Fix bugs in `_correct` & `_approximate` completers
```

Replaced `master` in the blob URL above with `4ec8790` - the commit before, and got the URL which now works.
</details>

**Reply by marlonrichert** [reply]
@L0laapk3 @Lockszmith-GH I created an issue to restore these instructions: #811
